votor/src/timer_manager/timers.rs

```diff
@@ -0,0 +1,243 @@
+    solana_ledger::leader_schedule_utils::last_of_consecutive_leader_slots,
+    solana_sdk::clock::Slot,
+    std::{
+        cmp::Reverse,
+        collections::{BinaryHeap, HashMap},
+        time::{Duration, Instant},
+    },
+};
+
+/// Encodes a basic state machine of the different stages involved in handling
+/// timeouts for a window of slots.
+enum TimerState {
+    /// Waiting for the DELTA_TIMEOUT stage.
+    WaitDeltaTimeout {
+        /// The slots in the window.  Must not be empty.
+        window: Vec<Slot>,
+        /// Time when this stage will end.
+        timeout: Instant,
+    },
+    /// Waiting for the DELTA_BLOCK stage.
+    WaitDeltaBlock {
+        /// The slots in the window.  Must not be empty.
+        window: Vec<Slot>,
+        /// Time when this stage will end.
+        timeout: Instant,
+    },
+    /// The state machine is done.
+    Done,
+}
+
+impl TimerState {
+    /// Creates a new instance of the state machine.
+    ///
+    /// Also returns the next time the timer should fire.
+    fn new(slot: Slot, delta_timeout: Duration) -> (Self, Instant) {
+        let mut window = (slot..=last_of_consecutive_leader_slots(slot)).collect::<Vec<_>>();
+        assert!(!window.is_empty());
+        window.reverse();
+        let timeout = Instant::now().checked_add(delta_timeout).unwrap();
+        (Self::WaitDeltaTimeout { window, timeout }, timeout)
+    }
+
+    /// Call to make progress on the state machine.
+    ///
+    /// Returns a potentially empty list of events that should be sent.
+    fn progress(&mut self, delta_block: Duration) -> Option<VotorEvent> {
+        let now = Instant::now();
+        match self {
+            Self::WaitDeltaTimeout { window, timeout } => {
+                assert!(!window.is_empty());
+                if &now < timeout {
+                    return None;
+                }
+                let slot = *window.last().unwrap();
+                let timeout = now.checked_add(delta_block).unwrap();
+                *self = Self::WaitDeltaBlock {
+                    window: window.to_owned(),
+                    timeout,
+                };
+                Some(VotorEvent::TimeoutCrashedLeader(slot))
+            }
+            Self::WaitDeltaBlock { window, timeout } => {
+                assert!(!window.is_empty());
+                if &now < timeout {
+                    return None;
+                }
+
+                let ret = Some(VotorEvent::Timeout(window.pop().unwrap()));
+                if window.is_empty() {
+                    *self = Self::Done;
+                } else {
+                    *timeout = now.checked_add(delta_block).unwrap();
+                }
+                ret
+            }
+            Self::Done => None,
+        }
+    }
+
+    /// When would this state machine next be able to make progress.
+    fn next_fire(&self) -> Option<Instant> {
+        match self {
+            Self::WaitDeltaTimeout { window: _, timeout } => Some(*timeout),
```

Comment:
nit you can combine the arms:
`Self::WaitDeltaTimeout { timeout, .. } | Self::WaitDeltaBlock { timeout, .. } => Some(*timeout),`